hardware/MQTT.cpp

```diff
@@ -945,3 +979,796 @@ void MQTT::SendSceneInfo(const uint64_t SceneIdx, const std::string & /*SceneNam
+		{
+			if (pSensor->state_topic.empty())
+			{
+				Log(LOG_ERROR, "A binary_sensor should have a state_topic!");
+				return;
+			}
+			InsertUpdateSwitch(pSensor);
+		}
+
+		//Check if we want to subscribe to this sensor
+		bool bDoSubscribe = false;
+
+		//Only add component_type = "sensor" for now
+		bDoSubscribe = ((pSensor->component_type == "sensor") || (pSensor->component_type == "binary_sensor") || (pSensor->component_type == "switch"));
+
+		if (bDoSubscribe)
+		{
+			if (!pSensor->availability_topic.empty())
+			{
+				SubscribeTopic(pSensor->availability_topic, pSensor->qos);
+			}
+			if (!pSensor->state_topic.empty())
+			{
+				SubscribeTopic(pSensor->state_topic, pSensor->qos);
+			}
+		}
+	}
+	catch (const std::exception &e)
+	{
+		Log(LOG_ERROR, "MQTT_Discovery: Error: %s!", e.what());
+		goto disovery_invaliddata;
+	}
+	return;
+disovery_invaliddata:
+	Log(LOG_ERROR, "MQTT_Discovery: Invalid data received!");
+}
+
+void MQTT::handle_auto_discovery_sensor_message(const struct mosquitto_message *message)
+{
+	std::string topic = message->topic;
+	std::string qMessage = std::string((char *)message->payload, (char *)message->payload + message->payloadlen);
+
+	if (qMessage.empty())
+		return;
+
+	for (auto& itt : m_discovered_sensors)
+	{
+		_tMQTTASensor *pSensor = &itt.second;
+
+		if (pSensor->state_topic == topic)
+		{
+			bool bIsJSON = false;
+			Json::Value root;
+			bool ret = ParseJSon(qMessage, root);
+			if (ret)
+			{
+				bIsJSON = root.isObject();
+			}
+			std::string szValue;
+			if (bIsJSON)
+			{
+				std::string value_template = pSensor->value_template;
+				if (value_template.empty())
+					continue; //we need a value template (could be "{\"state\":\"OFF\",\"brightness\":255,\"color\":{\"r\":255,\"g\":255,\"b\":255}}")
+
+				Json::Value key_value = root;
+
+				if (value_template.find('.') != std::string::npos)
+				{
+					std::vector<std::string> strarray;
+					StringSplit(value_template, ".", strarray);
+					if (strarray.size()>1)
+					{
+						for (auto ittKey : strarray)
+						{
+							if (!key_value[ittKey].empty())
+							{
+								if (key_value[ittKey].isObject())
+									key_value = key_value[ittKey];
+								else
+									value_template = ittKey;
+							}
+							else
+							{
+								//key not found!
+								continue;
+							}
+						}
+					}
+				}
+				
+
+				if (key_value[value_template].empty())
+					continue; // key not found
+				szValue = key_value[value_template].asString();
+			}
+			else
+			{
+				szValue = qMessage;
+			}
+			pSensor->last_value = szValue;
+			pSensor->last_received = mytime(nullptr);
+#ifdef _DEBUG
+			std::string szLogMessage = std_format("%s (value: %s", pSensor->name.c_str(), pSensor->last_value.c_str());
+			if (!pSensor->unit_of_measurement.empty())
+			{
+				szLogMessage += " " + pSensor->unit_of_measurement;
+			}
+			szLogMessage += ")";
+			Log(LOG_NORM, "MQTT received: %s", szLogMessage.c_str());
+#endif
```

Comment:
Instead of compile time _DEBUG, would it be nicer to use 'Debug(DEBUG_HARDWARE, ' so when a user wants to _debug_ what is going on, it can just be enabled through the Debug switches instead of re-compiling?

Reply:
Was thinking about this.... first I want to finish some work before refactoring this